File: android/app/src/main/java/com/noop/testcentre/DisplayPerformanceMonitor.kt
package com.noop.testcentre

import android.content.Context
import android.content.res.Configuration
import android.os.Debug
import android.view.Choreographer
import com.noop.analytics.DisplayMetrics
import com.noop.analytics.DisplayTrace

/**
 * The app-layer frame-time / hitch / memory monitor for the Display & Performance test mode (Kotlin twin
 * of the Swift DisplayPerformanceMonitor).
 *
 * CRITICAL CONTRACT: this monitor is a PERFORMANCE tool that must not itself cost performance. It runs
 * ONLY while the Display mode is active. start() registers a Choreographer.FrameCallback when the mode
 * toggles on; stop() removes it when the mode toggles off. There is NO perpetual frame callback: when the
 * mode is off, no callback is posted, no frame fires, and zero DISPLAY lines are emitted. A test pins
 * exactly that (DisplayPerformanceMonitorTest).
 *
 * The frame callback samples each frame's duration, counts hitches over a threshold, and emits a ROLLING
 * SUMMARY line once per window of frames (not per frame), so the trace is bounded. Every emitted line
 * rides the caller's sink (vm.ble.externalLog(line, DISPLAY), the redacting log sink). Touches only the
 * main looper Choreographer + the sink, so it is exercised on the plain JVM for the pure pieces; the
 * Choreographer wiring is the only Android-only part and is never run in a test.
 */
object DisplayPerformanceMonitor {

    /** A frame is a "hitch" when its duration exceeds this (ms), matching the Swift threshold. */
    const val HITCH_THRESHOLD_MS: Double = 33.0

    /** Emit one rolling summary per this many frames (~1 s at 60 Hz), matching the Swift window. */
    const val WINDOW_FRAMES: Int = 60

    private var running = false
    private var sink: ((String) -> Unit)? = null

    /**
     * CAPTURE-D (#797): a provider of the on-device DATA VOLUME, wired by the Test Centre screen to
     * [com.noop.data.WhoopRepository.dataVolumeSnapshot]. When set, [emitDataVolume] emits ONE upfront
     * `dataVolume` line so an import-driven-lag report shows the read-set the screens render over, not only
     * frame stats. null leaves the line unemitted (e.g. a test with no store). Mirrors the Swift
     * DisplayPerformanceMonitor.dataVolumeProvider.
     */
    var dataVolumeProvider: (suspend () -> com.noop.analytics.DataVolume?)? = null

    private var lastFrameNanos: Long = 0
    private val windowDurationsMs = ArrayList<Double>()
    private var windowHitches = 0
    private var windowWorstMs = 0.0
    private var memoryPeakMB = 0.0

    /** True only while the frame callback is registered. The test asserts this is false when off. */
    val isRunning: Boolean get() = running

    private val frameCallback = object : Choreographer.FrameCallback {
        override fun doFrame(frameTimeNanos: Long) {
            if (!running) return
            foldFrame(frameTimeNanos)
            // Re-post for the next frame ONLY while running, so removing the callback on stop() ends the
            // chain immediately (there is no standing registration to leak).
            Choreographer.getInstance().postFrameCallback(this)
        }
    }

    // Lifecycle (the ONLY place the frame callback is posted / removed) -------------------------------

    /**
     * Start the monitor: emit the device metrics immediately, sample memory, and begin the frame chain.
     * Idempotent. [context] reads the live Configuration for the metrics line; [emit] is the redacting
     * DISPLAY sink.
     */
    fun start(context: Context, emit: (String) -> Unit) {
        if (running) return
        running = true
        sink = emit
        resetWindow()
        lastFrameNanos = 0
        memoryPeakMB = 0.0
        sampleMemory()
        emit(DisplayTrace.deviceMetricsLine(readMetrics(context)))
        Choreographer.getInstance().postFrameCallback(frameCallback)
    }

    /**
     * CAPTURE-D (#797): read the on-device DATA VOLUME (via [dataVolumeProvider]) and emit ONE upfront
     * `dataVolume` line. Suspend because the read hits the store; the Test Centre screen calls it once from
     * a coroutine right after [start] (the monitor object itself owns no scope, mirroring how the Swift
     * monitor emits it in a Task wired by TestCentreView). A no-op when the monitor is off, the sink is
     * gone, or no provider is set. A null snapshot emits nothing (never a fabricated zero line).
     */
    suspend fun emitDataVolume() {
        if (!running) return
        val provider = dataVolumeProvider ?: return
        val volume = runCatching { provider() }.getOrNull() ?: return
        sink?.invoke(DisplayTrace.dataVolumeLine(volume))
    }

    /**
     * Stop the monitor: remove the frame callback (so nothing keeps firing), flush a final partial window
     * if it holds samples, and emit the memory high-water. After this no frame fires and no DISPLAY line is
     * emitted until the next start(). Idempotent.
     */
    fun stop() {
        if (!running) return
        running = false
        Choreographer.getInstance().removeFrameCallback(frameCallback)
        flushWindow()
        sink?.invoke(DisplayTrace.memoryHighWaterLine(memoryPeakMB))
        sink = null
    }

    // Frame sampling ---------------------------------------------------------------------------------

    private fun foldFrame(frameTimeNanos: Long) {
        val prev = lastFrameNanos
        lastFrameNanos = frameTimeNanos
        if (prev <= 0L) return // seed frame, no duration yet
        val durationMs = (frameTimeNanos - prev) / 1_000_000.0
        if (durationMs <= 0 || durationMs >= 5_000) return // ignore a backgrounding gap
        windowDurationsMs.add(durationMs)
        if (durationMs > HITCH_THRESHOLD_MS) windowHitches++
        if (durationMs > windowWorstMs) windowWorstMs = durationMs
        if (windowDurationsMs.size >= WINDOW_FRAMES) {
            sampleMemory()
            flushWindow()
        }
    }

    private fun flushWindow() {
        if (windowDurationsMs.isEmpty()) return
        val stats = windowStats(windowDurationsMs)
        sink?.invoke(
            DisplayTrace.frameSummaryLine(
                frames = windowDurationsMs.size,
                meanMs = stats.first, p95Ms = stats.second,
                hitches = windowHitches, worstMs = windowWorstMs,
                hitchThresholdMs = HITCH_THRESHOLD_MS,
            ),
        )
        resetWindow()
    }

    private fun resetWindow() {
        windowDurationsMs.clear()
        windowHitches = 0
        windowWorstMs = 0.0
    }

    /** Pure window stats: mean and p95 (nearest-rank) of the frame durations. Static and side-effect-free
     *  so a fixture can pin the exact summary numbers without a live frame callback. Mirrors the Swift
     *  windowStats. */
    fun windowStats(durationsMs: List<Double>): Pair<Double, Double> {
        if (durationsMs.isEmpty()) return 0.0 to 0.0
        val mean = durationsMs.sum() / durationsMs.size
        val sorted = durationsMs.sorted()
        // Nearest-rank p95 as a 0-based index: idx = ceil(0.95 * n), clamped to the last element. This
        // counts the values that fall BELOW the percentile, so at least 95% of frames are at or below the
        // returned value. For a hitch trace that matters: a window of 20 frames with one 5%-tail hitch
        // lands p95 ON the hitch (idx = ceil(0.95 * 20) = 19 = the worst of 20). Mirrors the Swift twin.
        val idx = Math.ceil(0.95 * sorted.size).toInt()
        val p95 = sorted[minOf(idx, sorted.size - 1)]
        return mean to p95
    }

    // Memory high-water ------------------------------------------------------------------------------

    private fun sampleMemory() {
        val mb = residentFootprintMB()
        if (mb > memoryPeakMB) memoryPeakMB = mb
    }

    /** Current resident footprint in MB via Debug total PSS (the "real memory" figure for the process).
     *  Kotlin twin of the Swift phys_footprint read. */
    private fun residentFootprintMB(): Double {
        val mi = Debug.MemoryInfo()
        Debug.getMemoryInfo(mi)
        return mi.totalPss / 1024.0 // totalPss is in KB
    }

    // Device metrics (read live, formatted by the pure DisplayTrace) ---------------------------------

    /** Read the current display environment into the pure DisplayMetrics carrier from the Android
     *  Configuration. Android has no true size class, so those are null (printed "n/a"); fontScale becomes
     *  the dynamicType label; densityDpi gives the backing scale; uiMode-night gives the theme. */
    fun readMetrics(context: Context): DisplayMetrics {
        val cfg: Configuration = context.resources.configuration
        val density = context.resources.displayMetrics.density.toDouble()
        val nightMode = cfg.uiMode and Configuration.UI_MODE_NIGHT_MASK
        val dark = nightMode == Configuration.UI_MODE_NIGHT_YES
        val portrait = cfg.orientation != Configuration.ORIENTATION_LANDSCAPE
        return DisplayMetrics(
            horizontalSizeClass = null, verticalSizeClass = null, // Android has no size class
            widthPt = cfg.screenWidthDp.toDouble(), heightPt = cfg.screenHeightDp.toDouble(),
            scale = density,
            safeTop = 0.0, safeBottom = 0.0, safeLeading = 0.0, safeTrailing = 0.0,
            dynamicType = fontScaleLabel(cfg.fontScale),
            orientation = if (portrait) "portrait" else "landscape",
            theme = if (dark) "dark" else "light",
        )
    }

    /** The Android Dynamic Type label is the font scale to two decimals ("1.00" / "1.30" / "1.50"), so a
     *  "text is clipped" report shows how large the user's text was. */
    internal fun fontScaleLabel(scale: Float): String = String.format(java.util.Locale.US, "%.2f", scale)
}

File: Strand/System/DisplayPerformanceMonitor.swift
import Foundation
import StrandAnalytics
#if os(iOS)
import UIKit
#endif
#if os(macOS)
import AppKit
import CoreVideo
#endif

// DisplayPerformanceMonitor.swift - the app-layer frame-time / hitch / memory monitor for the Display &
// Performance test mode (Test Centre).
//
// CRITICAL CONTRACT: this monitor is a PERFORMANCE tool that must not itself cost performance. It runs
// ONLY while the Display mode is active. start() is called when the mode toggles on (and on appear if it
// was already on); stop() is called when it toggles off. There is NO perpetual display link: when the
// mode is off, no CADisplayLink / CVDisplayLink exists, no frame callback fires, and zero `.display`
// lines are emitted. A test pins exactly that (DisplayPerformanceMonitorTests).
//
// The frame callback samples each frame's duration, counts hitches over a threshold, and emits a ROLLING
// SUMMARY line once per window of frames (not per frame), so the trace is bounded. Every emitted line
// rides LiveState.append(log:domain:.display), the single redacting sink. The screenshot + device-metrics
// readers live here too so the whole Display capture surface is one file.

/// Owns the live frame monitor for the Display mode. A single shared instance is started / stopped from
/// the Test Centre toggle. `@MainActor` because it touches the display link, the key window, and the
/// LiveState sink, all of which are main-thread concerns.
@MainActor
final class DisplayPerformanceMonitor {

    static let shared = DisplayPerformanceMonitor()
    private init() {}

    /// A frame is a "hitch" when its duration exceeds this (ms). 16.7 ms is a 60 Hz frame; 33 ms is a
    /// dropped frame at 60 Hz / a slow frame at 120 Hz, a sensible "the user felt that" threshold.
    static let hitchThresholdMs: Double = 33

    /// Emit one rolling summary per this many frames (~1 s at 60 Hz). Per-window, never per-frame, so the
    /// emission is a natural throttle and the strap log never floods.
    static let windowFrames = 60

    /// The sink that writes a tagged `.display` line. Set by the screen to LiveState.append(log:domain:);
    /// nil leaves the monitor inert (e.g. before the screen wired it, or in a test with no LiveState).
    var emit: ((String) -> Void)?

    /// CAPTURE-D (#797): reads the on-device DATA VOLUME straight from the STORE (never via the Repository /
    /// @Published view-models) so import-driven lag shows its READ-SET, not just frame stats. Wired by the
    /// screen alongside `emit`; nil leaves the dataVolume line unemitted (e.g. a test with no store). Async
    /// because the store is an actor; `start()` fires it on a detached task and emits the one line on return.
    var dataVolumeProvider: (() async -> DataVolume?)?

    private var running = false

    // Frame-time accumulation for the current window.
    private var lastFrameTimestamp: CFTimeInterval = 0
    private var windowDurationsMs: [Double] = []
    private var windowHitches = 0
    private var windowWorstMs: Double = 0

    /// Peak resident footprint (MB) seen since start(), emitted on stop() and periodically.
    private var memoryPeakMB: Double = 0

    #if os(iOS)
    private var displayLink: CADisplayLink?
    #endif
    #if os(macOS)
    private var cvLink: CVDisplayLink?
    #endif

    /// True only while a display link is live. The test asserts this is false when the mode is off.
    var isRunning: Bool { running }

    // MARK: - Lifecycle (the ONLY place a display link is created / destroyed)

    /// Start the monitor: create the platform display link and begin sampling. Idempotent (a second
    /// start is ignored). Emits one device-metrics line immediately so the report always has the layout
    /// environment even if the user reports before a frame window completes.
    func start() {
        guard !running else { return }
        running = true
        resetWindow()
        lastFrameTimestamp = 0
        memoryPeakMB = 0
        sampleMemory()
        emitDeviceMetrics()
        emitDataVolume()

        #if os(iOS)
        let link = CADisplayLink(target: self, selector: #selector(onFrame(_:)))
        link.add(to: .main, forMode: .common)
        displayLink = link
        #endif
        #if os(macOS)
        var link: CVDisplayLink?
        CVDisplayLinkCreateWithActiveCGDisplays(&link)
        if let link {
            // The CVDisplayLink callback fires on a private CV thread; hop to the main actor to fold the
            // sample in, so all monitor state stays main-isolated like the iOS path.
            CVDisplayLinkSetOutputHandler(link) { [weak self] _, inNow, _, _, _ in
                let nowSeconds = Double(inNow.pointee.hostTime) / Double(machTimebaseHz())
                Task { @MainActor in self?.foldFrame(timestamp: nowSeconds) }
                return kCVReturnSuccess
            }
            CVDisplayLinkStart(link)
            cvLink = link
        }
        #endif
    }

    /// Stop the monitor: tear the display link DOWN (so nothing keeps firing), flush a final partial
    /// window if it holds any samples, and emit the memory high-water. After this no callback fires and
    /// no `.display` line is emitted until the next start(). Idempotent.
    func stop() {
        guard running else { return }
        running = false

        #if os(iOS)
        displayLink?.invalidate()
        displayLink = nil
        #endif
        #if os(macOS)
        if let link = cvLink {
            CVDisplayLinkStop(link)
        }
        cvLink = nil
        #endif

        // Flush whatever the last partial window holds, then the high-water mark.
        flushWindow()
        emit?(DisplayTrace.memoryHighWaterLine(peakMB: memoryPeakMB))
    }

    // MARK: - Frame sampling

    #if os(iOS)
    @objc private func onFrame(_ link: CADisplayLink) {
        foldFrame(timestamp: link.timestamp)
    }
    #endif

    /// Fold one frame timestamp into the current window. The first frame seeds the previous-timestamp and
    /// is not counted (no duration yet). Once a window fills (`windowFrames`), it flushes a summary and
    /// resets. Pure arithmetic plus the periodic emit; no allocation per frame beyond the bounded buffer.
    private func foldFrame(timestamp: CFTimeInterval) {
        guard running else { return }
        defer { lastFrameTimestamp = timestamp }
        guard lastFrameTimestamp > 0 else { return }   // seed frame, no duration yet
        let durationMs = (timestamp - lastFrameTimestamp) * 1000.0
        guard durationMs > 0, durationMs < 5_000 else { return }  // ignore a backgrounding gap
        windowDurationsMs.append(durationMs)
        if durationMs > Self.hitchThresholdMs { windowHitches += 1 }
        if durationMs > windowWorstMs { windowWorstMs = durationMs }
        if windowDurationsMs.count >= Self.windowFrames {
            sampleMemory()
            flushWindow()
        }
    }

    /// Emit the rolling summary for the just-completed window and reset for the next one. No-op when the
    /// window is empty (nothing to summarise), so a stop() with no frames does not emit a zero line.
    private func flushWindow() {
        guard !windowDurationsMs.isEmpty else { return }
        let stats = DisplayPerformanceMonitor.windowStats(durationsMs: windowDurationsMs)
        emit?(DisplayTrace.frameSummaryLine(
            frames: windowDurationsMs.count,
            meanMs: stats.mean, p95Ms: stats.p95,
            hitches: windowHitches, worstMs: windowWorstMs,
            hitchThresholdMs: Self.hitchThresholdMs))
        resetWindow()
    }

    private func resetWindow() {
        windowDurationsMs.removeAll(keepingCapacity: true)
        windowHitches = 0
        windowWorstMs = 0
    }

    /// Pure window stats: mean and p95 of the frame durations. Static and side-effect-free so a fixture
    /// can pin the exact summary numbers without a live display link.
    static func windowStats(durationsMs: [Double]) -> (mean: Double, p95: Double) {
        guard !durationsMs.isEmpty else { return (0, 0) }
        let mean = durationsMs.reduce(0, +) / Double(durationsMs.count)
        let sorted = durationsMs.sorted()
        // Nearest-rank p95 as a 0-based index: idx = ceil(0.95 * n), clamped to the last element. This
        // counts the values that fall BELOW the percentile, so at least 95% of frames are at or below the
        // returned value. For a hitch trace that matters: a window of 20 frames with one 5%-tail hitch
        // lands p95 ON the hitch (idx = ceil(0.95 * 20) = 19 = the worst of 20), surfacing exactly the
        // slow frame the user is reporting rather than hiding it under the 95% of healthy frames.
        let idx = Int((0.95 * Double(sorted.count)).rounded(.up))
        let p95 = sorted[min(idx, sorted.count - 1)]
        return (mean, p95)
    }

    // MARK: - Memory high-water

    /// Sample the current resident footprint and raise the high-water mark. Reads phys_footprint via
    /// task_info (the same number Xcode's memory gauge shows); a read failure leaves the mark unchanged
    /// rather than fabricating a value.
    private func sampleMemory() {
        if let mb = Self.residentFootprintMB(), mb > memoryPeakMB { memoryPeakMB = mb }
    }

    /// Current resident footprint in MB via `task_info(TASK_VM_INFO).phys_footprint`, or nil on a read
    /// failure. The Apple-recommended "real memory" figure for a process.
    static func residentFootprintMB() -> Double? {
        var info = task_vm_info_data_t()
        var count = mach_msg_type_number_t(MemoryLayout<task_vm_info_data_t>.size / MemoryLayout<integer_t>.size)
        let kr = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                task_info(mach_task_self_, task_flavor_t(TASK_VM_INFO), $0, &count)
            }
        }
        guard kr == KERN_SUCCESS else { return nil }
        return Double(info.phys_footprint) / (1024.0 * 1024.0)
    }

    // MARK: - Device metrics (read live, formatted by the pure DisplayTrace)

    /// Read the live display metrics and emit the device-metrics line. Public so the screen can re-emit
    /// on a trait / orientation change while the mode is on.
    func emitDeviceMetrics() {
        emit?(DisplayTrace.deviceMetricsLine(Self.readMetrics()))
    }

    /// CAPTURE-D (#797): read the on-device DATA VOLUME from the store (via `dataVolumeProvider`) and emit
    /// ONE `dataVolume` line, so an import-driven-lag report shows the read-set the screens render over, not
    /// only frame stats. The provider is async (the store is an actor), so this fires a detached task and
    /// emits the line on the main actor on return. No provider wired (e.g. a test without a store) → no
    /// line, exactly as before. Side-effect-only; never touches the frame monitor.
    private func emitDataVolume() {
        guard let dataVolumeProvider, emit != nil else { return }
        Task { [weak self] in
            let volume = await dataVolumeProvider()
            await MainActor.run {
                guard let self, let volume else { return }
                self.emit?(DisplayTrace.dataVolumeLine(volume))
            }
        }
    }

    /// Read the current display environment into the pure DisplayMetrics carrier. iOS reads the key
    /// window's trait collection + safe-area insets + UIScreen; macOS reads the key NSWindow / NSScreen
    /// and degrades the iOS-only metrics (size class, Dynamic Type) to nil.
    static func readMetrics() -> DisplayMetrics {
        #if os(iOS)
        let scene = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first { $0.activationState == .foregroundActive }
            ?? UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }.first
        let window = scene?.keyWindow ?? scene?.windows.first
        let traits = window?.traitCollection ?? UITraitCollection.current
        let insets = window?.safeAreaInsets ?? .zero
        let bounds = window?.bounds.size ?? scene?.screen.bounds.size ?? .zero
        let scale = scene?.screen.scale ?? UIScreen.main.scale
        let portrait = (scene?.interfaceOrientation.isPortrait) ?? (bounds.height >= bounds.width)
        return DisplayMetrics(
            horizontalSizeClass: sizeClassName(traits.horizontalSizeClass),
            verticalSizeClass: sizeClassName(traits.verticalSizeClass),
            widthPt: Double(bounds.width), heightPt: Double(bounds.height), scale: Double(scale),
            safeTop: Double(insets.top), safeBottom: Double(insets.bottom),
            safeLeading: Double(insets.left), safeTrailing: Double(insets.right),
            dynamicType: contentSizeName(traits.preferredContentSizeCategory),
            orientation: portrait ? "portrait" : "landscape",
            theme: traits.userInterfaceStyle == .dark ? "dark" : "light")
        #elseif os(macOS)
        let window = NSApplication.shared.keyWindow ?? NSApplication.shared.windows.first
        let screen = window?.screen ?? NSScreen.main
        let size = window?.frame.size ?? screen?.frame.size ?? .zero
        let scale = screen?.backingScaleFactor ?? 0
        let dark = NSApp.effectiveAppearance.bestMatch(from: [.darkAqua, .aqua]) == .darkAqua
        return DisplayMetrics(
            horizontalSizeClass: nil, verticalSizeClass: nil,   // macOS has no size class
            widthPt: Double(size.width), heightPt: Double(size.height), scale: Double(scale),
            safeTop: 0, safeBottom: 0, safeLeading: 0, safeTrailing: 0,   // no notch insets on macOS
            dynamicType: nil,                                  // no Dynamic Type on macOS
            orientation: size.height >= size.width ? "portrait" : "landscape",
            theme: dark ? "dark" : "light")
        #else
        return DisplayMetrics(
            horizontalSizeClass: nil, verticalSizeClass: nil,
            widthPt: 0, heightPt: 0, scale: 0,
            safeTop: 0, safeBottom: 0, safeLeading: 0, safeTrailing: 0,
            dynamicType: nil, orientation: "unknown", theme: "light")
        #endif
    }

    #if os(iOS)
    private static func sizeClassName(_ c: UIUserInterfaceSizeClass) -> String? {
        switch c {
        case .compact: return "compact"
        case .regular: return "regular"
        default: return nil
        }
    }

    /// Short Dynamic Type label from the content-size category ("L" / "XXL" / "AX3"). The accessibility
    /// sizes map to "AX1"..."AX5", the standard sizes to their short forms, so the line is compact.
    private static func contentSizeName(_ c: UIContentSizeCategory) -> String {
        switch c {
        case .extraSmall: return "XS"
        case .small: return "S"
        case .medium: return "M"
        case .large: return "L"
        case .extraLarge: return "XL"
        case .extraExtraLarge: return "XXL"
        case .extraExtraExtraLarge: return "XXXL"
        case .accessibilityMedium: return "AX1"
        case .accessibilityLarge: return "AX2"
        case .accessibilityExtraLarge: return "AX3"
        case .accessibilityExtraExtraLarge: return "AX4"
        case .accessibilityExtraExtraExtraLarge: return "AX5"
        default: return "L"
        }
    }
    #endif
}

#if os(macOS)
/// The mach timebase, cached, so the CVDisplayLink host-time (mach ticks) converts to seconds. Pure read
/// of the timebase info; the ratio is constant for the process lifetime.
private func machTimebaseHz() -> UInt64 {
    var info = mach_timebase_info_data_t()
    mach_timebase_info(&info)
    // hostTime is in mach ticks; ticks * (numer/denom) = nanoseconds. ticks-per-second = 1e9 * denom/numer.
    guard info.numer != 0 else { return 1_000_000_000 }
    return UInt64(1_000_000_000) * UInt64(info.denom) / UInt64(info.numer)
}
#endif
